src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1569,6 +1569,352 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
+                                 XMMRegister ones, XMMRegister xtmp,
+                                 int vlen_enc) {
+  assert(VM_Version::supports_avx512vl(), "");
+  if (elem_bt == T_SHORT) {
+    evpmovdw(dst, dst, vlen_enc);
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    evpmovdb(dst, dst, vlen_enc);
+  }
+}
+
+void C2_MacroAssembler::vgather_subword_avx3(BasicType elem_bt, XMMRegister dst, Register base, XMMRegister offset,
+                                             Register idx_base, int idx_off, XMMRegister idx_vec, XMMRegister ones,
+                                             XMMRegister xtmp, KRegister gmask, int vlen_enc) {
+  if (elem_bt == T_SHORT) {
+    evmovdquq(idx_vec, Address(idx_base, idx_off, Address::times_4), vlen_enc);
+    if (offset != xnoreg) {
+      vpaddd(idx_vec, idx_vec, offset, vlen_enc);
+    }
+    // Normalize the indices to multiple of 2.
+    vpslld(xtmp, ones, 1, vlen_enc);
+    vpand(xtmp, idx_vec, xtmp, vlen_enc);
+    // Load double words from normalized indices.
+    evpgatherdd(dst, gmask, Address(base, xtmp, Address::times_2), vlen_enc);
+    // Compute bit level offset of actual short value with in each double word
+    // lane.
+    vpsrld(xtmp, ones, 31, vlen_enc);
+    vpand(xtmp, idx_vec, xtmp, vlen_enc);
+    vpslld(xtmp, xtmp, 4, vlen_enc);
+    // Move the short value at respective bit offset to lower 16 bits of each
+    // double word lane.
+    vpsrlvd(dst, dst, xtmp, vlen_enc);
+    // Pack double word vector into short vector.
+    vpackI2X(T_SHORT, dst, ones, xtmp, vlen_enc);
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    evmovdquq(idx_vec, Address(idx_base, idx_off, Address::times_4), vlen_enc);
+    if (offset != xnoreg) {
+      vpaddd(idx_vec, idx_vec, offset, vlen_enc);
+    }
+    // Normalize the indices to multiple of 4.
+    vpslld(xtmp, ones, 2, vlen_enc);
+    vpand(xtmp, idx_vec, xtmp, vlen_enc);
+    // Load double words from normalized indices.
+    evpgatherdd(dst, gmask, Address(base, xtmp, Address::times_1), vlen_enc);
+    // Compute bit level offset of actual byte value with in each double word
+    // lane.
+    vpsrld(xtmp, ones, 30, vlen_enc);
+    vpand(xtmp, idx_vec, xtmp, vlen_enc);
+    vpslld(xtmp, xtmp, 3, vlen_enc);
+    // Move the byte value at respective bit offset to lower 8 bits of each
+    // double word lane.
+    vpsrlvd(dst, dst, xtmp, vlen_enc);
+    // Pack double word vector into byte vector.
+    vpackI2X(T_BYTE, dst, ones, xtmp, vlen_enc);
+  }
+}
+
+void C2_MacroAssembler::vgather_subword_masked_avx3(BasicType elem_bt, XMMRegister dst, Register base, Register idx_base,
+                                                    Register offset, XMMRegister offset_vec, XMMRegister idx_vec,
+                                                    XMMRegister xtmp1, XMMRegister xtmp2, XMMRegister xtmp3, KRegister mask,
+                                                    KRegister gmask, int vlen_enc, int vlen) {
```

Comment:
Would you mind giving a quick summary of what the input registers are and what exactly this method does?
Why do we need to call `vgather_subword_avx3` so many times (`lane_count_subwords`)?

Reply:
Method gathers sub-words from gather indices using integral gather instructions, because of the lane size mismatch b/w int and sub-words algorithm makes multiple calls to vgather_subword_avx3.

Reply:
As a reviewer, I feel like I have to reverse engineer this now. I would really appreciate if there was a proper comment at the beginning, that tells me what is happening here. Maybe use some equation at the beginning, of what we want to acheive in the abstract, then explain why that does not work directly, and why you have to break it down into a loop, and then state the equation again in the loop form.